torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1214,10 +1214,8 @@ void ProcessGroupNCCL::heartbeatMonitor() {
       // We poll store to see if some ranks have flagged a timeout when
       // we haven't polled for `heartbeat_timeout` seconds and there haven't
       // any work added or removed for `watchdog_timeout` seconds.
-      if (computeDeltaMS(lastWorkListUpdateTime_, currentTime) >=
-              kWatchdogThreadSleepMillis &&
-          computeDeltaMS(lastTimePollStore, currentTime) >=
-              coordCheckIntervalMilSec_) {
+      if (computeDeltaMS(lastTimePollStore, currentTime) >=
+          coordCheckIntervalMilSec_) {
```

Comment:
In your PR desc it sounded like we already moved store check to heartbeat thread, but it is still in watchdog thread in this version. I believe if we remove last work check we may risk polling the store more frequently and overwhelming it. (At least that's why it was put there)

Reply:
When the code gets here, the watchdog has already hung (heartbeat thread's killing kicking in) -- I am not sure why we would still care about the last time watchdog thread checked the queue. The frequency we poll the store can be simply controlled by `coordCheckIntervalMilSec_`.

Reply:
This is the move that moves the store check from watchdog to heartbeat thread: https://github.com/pytorch/pytorch/pull/117168
My understanding is that if the store check is still in watchdog thread, then yes, we can use `lastWorkListUpdateTime_ `. But after the move, it can be decoupled.

Reply:
> When the code gets here, the watchdog has already hung (heartbeat thread's killing kicking in)

this code is actually still inside the 'while(true)' block which is the 'armed' state for the heartbeat monitor, not the 'killing' state.  We are doing the check precisely becuase we aren't already killing, and want to know if other ranks are killing.

we want to avoid making the check every interval, bc its a lot of traffic on the TCPStore.  The optimization (@zdevito's idea) is that if the local watchdog itself is making any progress, then its safe to assume we aren't globally hung.  If we're globally hung, soon enough our local watchdog will also stop progressing.

So even after moving the store check, this isn't decoupled (we're still using it for the same purpose)

Reply:
> this code is actually still inside the 'while(true)' block which is the 'armed' state for the heartbeat monitor, not the 'killing' state
You are right. Thanks for calling it out. 

> The optimization (@zdevito's idea) is that if the local watchdog itself is making any progress, then its safe to assume we aren't globally hung. 
I like that idea. Even further:
when my peer's watchdog fires a timeout, it is expected my local watchdog will fire a timeout soon (bc every rank has the same timeout value). When would my watchdog fail to fire a timeout? My watchdog hangs. Thus, would it make sense to move this poll to a point where there is no heartbeat? That would give us a least frequent polling too. It would also mean this global polling is informational only, because the heartbeat thread would start the killing anyway.

Reply:
Yes. I think previously we incremented heartbeat outside work loop, so progress was an extra signal.  I recently changed to increment heartbeat inside loop, so no heartbeat should == no progress. So it should be safe to remove work progress counter and rely on timeout. If I'm not forgetting something...

Reply:
Thanks for the confirmation! I will rework this PR to move the poll too.

Reply:
Oh I might have forgotten something.

If the work list stays empty for a while, we still have to check the store. That still requires the extra timer.

Reply:
If the work queue stays empty forever, then this rank can run to the end and exit, we can let it do so.

If the work queue stays empty for a while and then enqueues something later, this rank would eventually timeout if peers are gone.

Reply:
the point of using TCPStore for coordination is to ensure everyone dumps within a reasonable time limit.  We don't want to wait for "enqueues something later" probably?